// fs-json-store/lib/private/fs-impl/mem-fs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var memfs_1 = require("memfs");
var util_1 = require("util");
// keep definition on file top
exports.NAME = "internal.memfs";
function volume() {
    var vol = new memfs_1.Volume();
    var impl = memfs_1.createFsFromVolume(vol);
    return {
        _impl: impl,
        _name: exports.NAME,
        chmod: util_1.promisify(impl.chmod),
        chown: util_1.promisify(impl.chown),
        mkdir: util_1.promisify(impl.mkdir),
        open: util_1.promisify(impl.open),
        close: util_1.promisify(impl.close),
        fsync: util_1.promisify(impl.fsync),
        readFile: util_1.promisify(impl.readFile),
        realpath: util_1.promisify(impl.realpath),
        rename: util_1.promisify(impl.rename),
        stat: util_1.promisify(impl.stat),
        writeFile: util_1.promisify(impl.writeFile),
        writeFileAtomic: util_1.promisify(impl.writeFile),
        unlink: util_1.promisify(impl.unlink),
    };
}
exports.volume = volume;
exports.fs = volume();
